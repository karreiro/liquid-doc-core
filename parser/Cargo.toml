[package]
name = "liquid_doc_parser"
version = "0.1.0"
edition = "2021"

[lib]
crate-type = ["rlib"]

[dependencies]
serde.workspace = true
serde_json.workspace = true
pest = "2.7"
pest_derive = "2.7"

[dev-dependencies]
insta.workspace = true
paste = "1.0"
pretty_assertions.workspace = true
