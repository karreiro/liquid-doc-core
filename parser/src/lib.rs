mod ast;
mod parser;

// Re-export all AST types
pub use ast::*;

// Re-export parser functionality
pub use parser::{parse_liquid_string, visit, LiquidParser};
