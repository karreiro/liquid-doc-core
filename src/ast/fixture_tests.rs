use crate::parser::parse_liquid_string;
use std::fs;

// Include the generated constants and macros from the build script
include!(concat!(env!("OUT_DIR"), "/test_files.rs"));

// Re-export the generated constants for easy access from other modules
pub use {FIXTURE_NAMES, FIXTURE_PATHS};

/// Macro to generate a test function for a specific fixture file
macro_rules! generate_fixture_test {
    ($fixture_name:expr) => {
        paste::paste! {
            #[test]
            fn [<test_fixture_ $fixture_name>]() {
                let fixture_path = concat!("web/fixtures/", $fixture_name, ".liquid");
                let content = fs::read_to_string(fixture_path)
                    .unwrap_or_else(|_| panic!("Failed to read fixture file: {}", fixture_path));

                // Test that the file can be parsed without errors
                let result = parse_liquid_string(&content, None);
                assert!(result.is_some(), "Failed to parse fixture: {}", $fixture_name);

                let ast = result.unwrap();

                // Basic validation that we got some nodes
                assert!(!ast.nodes.is_empty(), "No nodes found in fixture: {}", $fixture_name);

                // You can add more specific assertions here based on what you expect
                println!("‚úÖ Successfully parsed {} with {} nodes", $fixture_name, ast.nodes.len());
            }
        }
    };
}

/// Macro to generate a benchmarking test for a specific fixture file
macro_rules! generate_fixture_benchmark {
    ($fixture_name:expr) => {
        paste::paste! {
            #[cfg(test)]
            fn [<bench_fixture_ $fixture_name>]() {
                let fixture_path = concat!("web/fixtures/", $fixture_name, ".liquid");
                let content = fs::read_to_string(fixture_path)
                    .unwrap_or_else(|_| panic!("Failed to read fixture file: {}", fixture_path));

                let start = std::time::Instant::now();
                let iterations = 1000;

                for _ in 0..iterations {
                    let _result = parse_liquid_string(&content, None);
                }

                let duration = start.elapsed();
                let avg_time = duration.as_nanos() as f64 / iterations as f64;

                println!("üìä Benchmark {}: {:.2} ns/parse", $fixture_name, avg_time);
            }
        }
    };
}

/// Macro to generate a JSON serialization test for a specific fixture file
macro_rules! generate_fixture_json_test {
    ($fixture_name:expr) => {
        paste::paste! {
            #[test]
            fn [<test_fixture_ $fixture_name _json_roundtrip>]() {
                let fixture_path = concat!("web/fixtures/", $fixture_name, ".liquid");
                let content = fs::read_to_string(fixture_path)
                    .unwrap_or_else(|_| panic!("Failed to read fixture file: {}", fixture_path));

                let ast = parse_liquid_string(&content, None)
                    .expect(&format!("Failed to parse fixture: {}", $fixture_name));

                // Test JSON serialization
                let json = serde_json::to_string(&ast)
                    .expect(&format!("Failed to serialize fixture {} to JSON", $fixture_name));

                // Test JSON deserialization
                let _deserialized: crate::ast::LiquidAST = serde_json::from_str(&json)
                    .expect(&format!("Failed to deserialize fixture {} from JSON", $fixture_name));

                println!("‚úÖ JSON roundtrip successful for {}", $fixture_name);
            }
        }
    };
}

// Generate all fixture tests using the macro from build.rs
for_each_fixture_file!(generate_fixture_test);

// Generate all JSON roundtrip tests
for_each_fixture_file!(generate_fixture_json_test);

// Generate all benchmark tests
for_each_fixture_file!(generate_fixture_benchmark);

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_fixture_file_names_macro() {
        let names = fixture_file_names!();
        assert!(!names.is_empty(), "Should have fixture file names");

        // Check that we have all 9 fixtures
        assert_eq!(names.len(), 9, "Should have 9 fixture files");

        // Check that all names start with "fixture"
        for name in names {
            assert!(
                name.starts_with("fixture"),
                "All fixture names should start with 'fixture': {}",
                name
            );
        }

        println!("üìã Available fixtures: {:?}", names);
    }

    #[test]
    fn test_fixture_file_paths_macro() {
        let paths = fixture_file_paths!();
        assert!(!paths.is_empty(), "Should have fixture file paths");

        // Check that all paths exist
        for path in paths {
            assert!(
                fs::metadata(path).is_ok(),
                "Fixture file should exist: {}",
                path
            );
        }

        println!("üìÅ Fixture paths verified: {:?}", paths);
    }

    #[test]
    fn test_all_fixtures_can_be_parsed() {
        let paths = fixture_file_paths!();
        let mut successful_parses = 0;

        for path in paths {
            let content = fs::read_to_string(path)
                .unwrap_or_else(|_| panic!("Failed to read fixture file: {}", path));

            if let Some(ast) = parse_liquid_string(&content, None) {
                successful_parses += 1;
                println!("‚úÖ Parsed {}: {} nodes", path, ast.nodes.len());
            } else {
                println!("‚ùå Failed to parse {}", path);
            }
        }

        assert_eq!(
            successful_parses,
            paths.len(),
            "All fixtures should parse successfully"
        );
    }

    /// Helper function to run benchmarks on all fixtures
    #[test]
    fn benchmark_all_fixtures() {
        println!("\nüöÄ Running benchmarks on all fixtures...");

        // Use the procedurally generated benchmark functions
        bench_fixture_fixture1();
        bench_fixture_fixture2();
        bench_fixture_fixture3();
        bench_fixture_fixture4();
        bench_fixture_fixture5();
        bench_fixture_fixture6();
        bench_fixture_fixture7();
        bench_fixture_fixture8();
        bench_fixture_fixture9();
    }
}
