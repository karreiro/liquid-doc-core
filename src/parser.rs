use crate::ast::{LiquidDocParamNode, *};
use pest::Parser;
use pest_derive::Parser;
use wasm_bindgen::{prelude::wasm_bindgen, JsValue};

#[derive(Parser)]
#[grammar = "liquid.pest"]
pub struct LiquidParser;

pub fn visit(
    ast: &mut LiquidAST,
    pair: pest::iterators::Pair<Rule>,
    position_offset: Option<usize>,
) {
    match pair.as_rule() {
        Rule::Document => {
            // The Document rule is the root of the AST, so we can just ignore it
            for inner_pair in pair.into_inner() {
                visit(ast, inner_pair, position_offset);
            }
        }
        Rule::ImplicitDescription => {
            let node = LiquidDocDescriptionNode::implicit(&pair, position_offset);
            if !node.content.is_empty() {
                ast.add_node(LiquidNode::LiquidDocDescriptionNode(node));
            }
        }
        Rule::LiquidDocNode => {
            let mut content = pair.into_inner();
            let next = content.next().unwrap();
            match next.as_rule() {
                Rule::paramNode => {
                    let node = LiquidDocParamNode::new(&next, position_offset);

                    ast.add_node(LiquidNode::LiquidDocParamNode(node));
                }
                Rule::exampleNode => {
                    let node = LiquidDocExampleNode::from_pair(&next, position_offset);
                    ast.add_node(LiquidNode::LiquidDocExampleNode(node));
                }
                Rule::descriptionNode => {
                    let node = LiquidDocDescriptionNode::explicit(&next, position_offset);
                    ast.add_node(LiquidNode::LiquidDocDescriptionNode(node));
                }
                Rule::promptNode => {
                    // Process prompt node
                    unimplemented!("Prompt nodes are not yet implemented");
                }
                Rule::fallbackNode => {
                    // Process fallback node
                    unimplemented!("Fallback nodes are not yet implemented");
                }
                _ => unreachable!("Unexpected rule in LiquidDocNode: {:?}", next.as_rule()),
            }
        }
        Rule::TextNode => {
            let text_node = TextNode::from_pair(&pair, position_offset);
            if !text_node.is_empty() {
                ast.add_node(LiquidNode::TextNode(text_node));
            }
        }
        _ => unimplemented!("Handle rule: {:?}", pair.as_rule()),
    }
}

pub(crate) fn parse_liquid_string(
    input: &str,
    position_offset: Option<usize>,
) -> Option<LiquidAST> {
    let text = LiquidParser::parse(Rule::Document, input)
        .map_err(|e| println!("Parsing error: {}", e))
        .ok()?;

    let mut ast = LiquidAST::new();
    for pair in text {
        visit(&mut ast, pair, position_offset);
    }

    Some(ast)
}

#[wasm_bindgen]
pub fn parse_liquid(input: &str) -> JsValue {
    serde_wasm_bindgen::to_value(&parse_liquid_string(input, None))
        .expect("The LiquidAst was not in the correct format")
}
