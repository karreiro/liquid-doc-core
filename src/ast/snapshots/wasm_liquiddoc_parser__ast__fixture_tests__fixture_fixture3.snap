---
source: src/ast/fixture_tests.rs
expression: "@example\nFirst Example\n@example\nSecond Example"
---
[
  {
    "type": "LiquidDocExampleNode",
    "name": "example",
    "position": {
      "start": 10,
      "end": 33
    },
    "source": "@example\nFirst Example\n",
    "content": {
      "type": "TextNode",
      "value": "@example\nFirst Example\n",
      "position": {
        "start": 10,
        "end": 33
      },
      "source": "@example\nFirst Example\n"
    },
    "isInline": true
  },
  {
    "type": "LiquidDocExampleNode",
    "name": "example",
    "position": {
      "start": 33,
      "end": 56
    },
    "source": "@example\nSecond Example",
    "content": {
      "type": "TextNode",
      "value": "@example\nSecond Example",
      "position": {
        "start": 33,
        "end": 56
      },
      "source": "@example\nSecond Example"
    },
    "isInline": true
  }
]
